pkg/datapath/linux/node_ids.go

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package linux
+
+import (
+	"net"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+)
+
+const (
+	minNodeID = 1
+	maxNodeID = idpool.ID(^uint16(0))
+)
+
+// AllocateNodeID allocates a new ID for the given node (by IP) if one wasn't
+// already assigned.
+func (n *linuxNodeHandler) AllocateNodeID(nodeIP net.IP) uint16 {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	return n.allocateNodeIDLocked(nodeIP)
+}
+
+func (n *linuxNodeHandler) allocateNodeIDLocked(nodeIP net.IP) uint16 {
+	if len(nodeIP) == 0 || nodeIP.IsUnspecified() {
+		log.Error("Attempt to allocate a node ID for an empty node IP address")
```

Comment:
```suggestion
		log.Error("cannot allocate a node ID for an empty node IP address")
```
nit: this message could be more descriptive of a failure

Reply:
I think the weird error message is because I wouldn't consider this a failure: we shouldn't allocate node IDs for empty node IPs. I've added this just so we know if this ever happens and we can check if we missed some weird code path (e.g., this was happening in a previous version because the proxy would call this function with an empty IP).

Instead of changing the message, I think I'll make it a warning (seem more accurate as it's likely nothing bad happened, but we want to know about it) and I'll ensure the CI fails on that warning.